File: hphp/runtime/ext/math/ext_math.php
<?hh

/**
 * Get value of pi
 *
 * @return float - The value of pi as a float
 */
function pi(): float {
  return M_PI;
}

/**
 * Find lowest value
 *
 * @param mixed $values - An array containing the values or list each value
 *                        individually
 *
 * @return mixed - The numerically lowest of the parameter values.
 */
function min(mixed $values, ...): mixed {
  $ret = null;
  if(!is_array($values)) {
    $values = func_get_args();
  }
  if($values) {
    $ret = array_shift($values);
    foreach($values as $value) {
      if($value < $ret) {
        $ret = $value;
      }
    }
  }
  return $ret;
}

/**
 * Find highest value
 *
 * @param mixed $values - An array containing the values or list each value
 *                        individually
 *
 * @return mixed - The numerically highest of the parameter values.
 */
function max(mixed $values, ...): mixed {
  $ret = null;
  if(!is_array($values)) {
    $values = func_get_args();
  }
  if($values) {
    $ret = array_shift($values);
    foreach($values as $value) {
      if($value > $ret) {
        $ret = $value;
      }
    }
  }
  return $ret;
}

/**
 * Absolute value
 *
 * @param mixed $number - The numeric value to process
 * 
 * @return mixed - The absolute value of $number. If the argument $number is of
 *                 type float, the return type is also float, otherwise it is
 *                 integer
 */
<<__Native>>
function abs(mixed $number): mixed;
/**
 * Finds whether a value is a legal finite number
 *
 * @param float $value - The value to check
 *
 * @return bool - TRUE if val is a legal finite number within the allowed range for a
 *                PHP float on this platform, else FALSE.
 */
<<__Native>>
function is_finite(float $value): bool;

/**
 * Finds whether a value is infinite
 *
 * @param float $value - The value to check
 *
 * @return bool - TRUE if val is infinite, else FALSE.
 */
<<__Native>>
function is_infinite(float $value): bool;

/**
 * Finds whether a value is not a number
 *
 * @param float $value - The value to check
 *
 * @return bool - Returns TRUE if val is 'not a number', else FALSE.
 */
<<__Native>>
function is_nan(float $value): bool;

/**
 * Round fractions up
 * 
 * @param float $value - The value to round
 *
 * @return float - $value rounded up to the next highest integer.
 */
<<__Native>>
function ceil(float $value): float;

/**
 * Round fractions down
 * 
 * @param float $value - The value to round
 *
 * @return float - $value rounded to the next lowest integer.
 */
<<__Native>>
function floor(float $value): float;

/**
 * Rounds a float
 * 
 * @param float $value   - The value to round
 * @param int $precision - The optional number of decimal digits to round to.
 * @param int $mode      - Use one of the PHP_ROUND_* constants to specify the
 *                         mode in which rounding occurs.
 *
 * @return float - The rounded value
 */
<<__Native>>
function round(float $value, int $precision = 0,
               int $mode = PHP_ROUND_HALF_UP): float;

/**
 * Converts the number in degrees to the radian equivalent
 * 
 * @param float $number - Angular value in degrees
 * 
 * @return float - The radian equivalent of $number
 */
function deg2rad(float $number): float {
  return $number / 180.0 * M_PI;
}

/**
 * Converts the number in radians to the degrees equivalent
 * 
 * @param float $number - Angular value in radians
 * 
 * @return float - The degrees equivalent of $number
 */
function rad2deg(float $number): float {
  return $number / M_PI * 180.0;
}

/**
 * Decimal to binary
 * 
 * @param int $number - Decimal value to convert
 *
 * @return string - Binary string representation of $number
 */
<<__Native>>
function decbin(int $number): string;

/**
 * Decimal to hexadecimal
 * 
 * @param int $number - Decimal value to convert
 *
 * @return string - Hexadecimal string representation of $number
 */
<<__Native>>
function dechex(int $number): string;

/**
 * Decimal to octal
 * 
 * @param int $number - Decimal value to convert
 *
 * @return string - Octal string representation of $number
 */
<<__Native>>
function decoct(int $number): string;

/**
 * Binary to decimal
 *
 * @param string $binary_string - The binary string to convert
 *
 * @return mixed - The decimal value of $binary_string
 */
<<__Native>>
function bindec(mixed $binary_string): mixed;

/**
 * Hexadecimal to decimal
 *
 * @param string $hex_string - The hexadecimal string to convert
 *
 * @return mixed - The decimal value of $hex_string
 */
<<__Native>>
function hexdec(mixed $hex_string): mixed;

/**
 * Octal to decimal
 *
 * @param string $octal_string - The octal string to convert
 *
 * @return mixed - The decimal value of $octal_string
 */
<<__Native>>
function octdec(mixed $octal_string): mixed;

/**
 * Convert a number between arbitrary bases
 *
 * @param string $number - The number to convert
 * @param int $frombase  - The base $number is in
 * @param int $tobase    - The base to convert $number to
 *
 * @return string - $number converted to base $tobase
 */
<<__Native>>
function base_convert(mixed $number, int $frombase, int $tobase): string;

/**
 * Exponential expression
 *
 * @param mixed $base - The base to use
 * @param mixed $exp  - The exponent
 *
 * @return mixed - $base raised to the power of $exp. If both arguments are
 *                 non-negative integers and the result can be represented as
 *                 an integer, the result will be returned with integer type,
 *                 otherwise it will be returned as a float.
 */
<<__Native>>
function pow(mixed $base, mixed $exp): mixed;

/**
 * Calculates the exponent of e
 *
 * @param float $arg - The argument to process
 *
 * @return float - e raised to the power $arg
 */
<<__Native>>
function exp(float $arg): float;

/**
 * Returns exp(number) - 1, computed in a way that is accurate even when the
 * value of number is close to zero.
 *
 * @param float $arg - The argument to process
 *
 * @return float - e to the power of $arg minus one.
 */
<<__Native>>
function expm1(float $arg): float;

/**
 * Base-10 logarithm
 *
 * @param float $arg - The argument to process
 *
 * @return float - The base-10 logarithm of $arg
 */
<<__Native>>
function log10(float $arg): float;

/**
 * Returns log(1 + number), computed in a way that is accurate even when the
 * value of number is close to zero.
 *
 * @param float $number - The argument to process
 *
 * @return float - log(1 + $number)
 */
<<__Native>>
function log1p(float $number): float;

/**
 * Natural logarithm
 *
 * @param float $arg  - The value to calculate the logarithm for
 * @param float $base - The optional logarithmic base to use
 *
 * @return float - The logarithm of $arg to $base, if given, or the natural
 *                 logarithm.
 */
<<__Native>>
function log(float $number, float $base = 0.0): float;

/**
 * cos
 *
 * @param float $arg - The argument to process
 *
 * @return float - The cos of $arg
 */
<<__Native>>
function cos(float $arg): float;

/**
 * cosh
 *
 * @param float $arg - The argument to process
 *
 * @return float - The cosh of $arg
 */
<<__Native>>
function cosh(float $arg): float;

/**
 * sin
 *
 * @param float $arg - The argument to process
 *
 * @return float - The sin of $arg
 */
<<__Native>>
function sin(float $arg): float;

/**
 * sinh
 *
 * @param float $arg - The argument to process
 *
 * @return float - The sinh of $arg
 */
<<__Native>>
function sinh(float $arg): float;

/**
 * tan
 *
 * @param float $arg - The argument to process
 *
 * @return float - The tan of $arg
 */
<<__Native>>
function tan(float $arg): float;

/**
 * tanh
 *
 * @param float $arg - The argument to process
 *
 * @return float - The tanh of $arg
 */
<<__Native>>
function tanh(float $arg): float;

/**
 * acos
 *
 * @param float $arg - The argument to process
 *
 * @return float - The acos of $arg
 */
<<__Native>>
function acos(float $arg): float;

/**
 * acosh
 *
 * @param float $arg - The argument to process
 *
 * @return float - The acosh of $arg
 */
<<__Native>>
function acosh(float $arg): float;

/**
 * asin
 *
 * @param float $arg - The argument to process
 *
 * @return float - The asin of $arg
 */
<<__Native>>
function asin(float $arg): float;

/**
 * asinh
 *
 * @param float $arg - The argument to process
 *
 * @return float - The asinh of $arg
 */
<<__Native>>
function asinh(float $arg): float;

/**
 * atan
 *
 * @param float $arg - The argument to process
 *
 * @return float - The atan of $arg
 */
<<__Native>>
function atan(float $arg): float;

/**
 * atanh
 *
 * @param float $arg - The argument to process
 *
 * @return float - The atanh of $arg
 */
<<__Native>>
function atanh(float $arg): float;

/**
 * atanh2
 *
 * @param float $y - Dividend parameter
 * @param float $x - Divisor parameter
 *
 * @return float - The arc tangent of $y/$x 
 */
<<__Native>>
function atan2(float $y, float $x): float;

/**
 * Calculate the length of the hypotenuse of a right-angle triangle
 *
 * @param float $x - Length of first side
 * @param float $y - Length of second side
 *
 * @return float - Calculated length of the hypotenuse
 */
function hypot(float $x, float $y): float {
  return sqrt($x * $x + $y * $y);
}

/**
 * Returns the floating point remainder of the division of the arguments
 *
 * @param float $x - The dividend
 * @param float $y - The divisor
 *
 * @return float - The floating point remainder of $x/$y
 */
<<__Native>>
function fmod(float $x, float $y): float;

/**
 * Square root
 *
 * @param float $arg - The argument to process
 *
 * @return The square root of $arg or NAN for negative numbers.
 */
<<__Native>>
function sqrt(float $arg): float;

/**
 * Show largest possible random value
 * 
 * @return int - The largest possible random value returned by rand()
 */
<<__Native>>
function getrandmax(): int;

/**
 * Seed the random number generator
 *
 * @param int $seed - Optional seed value
 */
<<__Native>>
function srand(?int $seed = null): void;

/**
 * Generate a random integer
 *
 * @param int $min - The lowest value to return
 * @param int $max - The highest value to return
 *
 * @return int - A pseudo random value between $min and $max
 */
<<__Native>>
function rand(int $min = 0, int $max = -1): int;

/**
 * Show largest possible random value
 * 
 * @return int - The largest possible random value returned by mt_rand()
 */
<<__Native>>
function mt_getrandmax(): int;

/**
 * Seed the better random number generator
 *
 * @param int $seed - Optional seed value
 */
<<__Native>>
function mt_srand(?int $seed = null): void;

/**
 * Generate a better random value
 *
 * @param int $min - The lowest value to return
 * @param int $max - The highest value to return
 *
 * @return int - A pseudo random value between $min and $max
 */
<<__Native>>
function mt_rand(int $min = 0, int $max = -1): int;

/**
 * Combined linear congruential generator
 *
 * @return float - A pseudo random float value in the range of (0, 1)
 */
<<__Native>>
function lcg_value(): float;

File: hphp/runtime/ext/math/ext_math.cpp
/*
   +----------------------------------------------------------------------+
   | HipHop for PHP                                                       |
   +----------------------------------------------------------------------+
   | Copyright (c) 2010-2013 Facebook, Inc. (http://www.facebook.com)     |
   | Copyright (c) 1997-2010 The PHP Group                                |
   +----------------------------------------------------------------------+
   | This source file is subject to version 3.01 of the PHP license,      |
   | that is bundled with this package in the file LICENSE, and is        |
   | available through the world-wide-web at the following url:           |
   | http://www.php.net/license/3_01.txt                                  |
   | If you did not receive a copy of the PHP license and are unable to   |
   | obtain it through the world-wide-web, please send a note to          |
   | license@php.net so we can mail you a copy immediately.               |
   +----------------------------------------------------------------------+
*/

#include "hphp/runtime/ext/math/ext_math.h"
#include "hphp/runtime/vm/runtime.h"
#include "hphp/runtime/base/zend-math.h"
#include "hphp/runtime/base/zend-multiply.h"
#include "hphp/runtime/base/zend-string.h"
#include <math.h>

const double k_M_PI = 3.1415926535898;

#if defined(__APPLE__)
#ifndef isnan
#define isnan(x)  \
  ( sizeof (x) == sizeof(float )  ? __inline_isnanf((float)(x)) \
  : sizeof (x) == sizeof(double)  ? __inline_isnand((double)(x))  \
  : __inline_isnanl ((long double)(x)))
#endif

#ifndef isinf
#define isinf(x)  \
  ( sizeof (x) == sizeof(float )  ? __inline_isinff((float)(x)) \
  : sizeof (x) == sizeof(double)  ? __inline_isinfd((double)(x))  \
  : __inline_isinfl ((long double)(x)))
#endif
#endif

namespace HPHP {
///////////////////////////////////////////////////////////////////////////////

const int64_t k_PHP_ROUND_HALF_UP =   PHP_ROUND_HALF_UP;
const int64_t k_PHP_ROUND_HALF_DOWN = PHP_ROUND_HALF_DOWN;
const int64_t k_PHP_ROUND_HALF_EVEN = PHP_ROUND_HALF_EVEN;
const int64_t k_PHP_ROUND_HALF_ODD =  PHP_ROUND_HALF_ODD;

/* Logic based on zend_operators.c::convert_scalar_to_number() */
static DataType zend_convert_scalar_to_number(CVarRef num,
                                              int64_t &ival,
                                              double &dval) {
  DataType dt = num.toNumeric(ival, dval, true);
  if ((dt == KindOfDouble) || (dt == KindOfInt64)) {
    return dt;
  }

  if (num.isBoolean() || num.isNull() || num.isObject() || num.isResource() ||
      num.isString()) {
    ival = num.toInt64();
    return KindOfInt64;
  }

  // Fallback, callers will handle this as an error
  ival = 0;
  dval = 0.0;
  return num.getType();
}

Variant HHVM_FUNCTION(abs, CVarRef number) {
  int64_t ival;
  double dval;
  DataType k = zend_convert_scalar_to_number(number, ival, dval);
  if (k == KindOfDouble) {
    return fabs(dval);
  } else if (k == KindOfInt64) {
    return ival >= 0 ? ival : -ival;
  } else {
    return false;
  }
}

bool HHVM_FUNCTION(is_finite, double val) { return finite(val);}
bool HHVM_FUNCTION(is_infinite, double val) { return isinf(val);}
bool HHVM_FUNCTION(is_nan, double val) { return isnan(val);}

Variant HHVM_FUNCTION(ceil, CVarRef number) {
  int64_t ival;
  double dval;
  DataType k = zend_convert_scalar_to_number(number, ival, dval);
  if (k == KindOfInt64) {
    dval = (double)ival;
  } else if (k != KindOfDouble) {
    return false;
  }
  return ceil(dval);
}

Variant HHVM_FUNCTION(floor, CVarRef number) {
  int64_t ival;
  double dval;
  DataType k = zend_convert_scalar_to_number(number, ival, dval);
  if (k == KindOfInt64) {
    dval = (double)ival;
  } else if (k != KindOfDouble) {
    return false;
  }
  return floor(dval);
}

Variant HHVM_FUNCTION(round, CVarRef val, int64_t precision /* = 0 */,
                int64_t mode /* = PHP_ROUND_HALF_UP */) {
  int64_t ival;
  double dval;
  DataType k = zend_convert_scalar_to_number(val, ival, dval);
  if (k == KindOfInt64) {
    if (precision >= 0) {
     return (double)ival;
    } else {
      dval = ival;
    }
  } else if (k != KindOfDouble) {
    return false;
  }
  dval = php_math_round(dval, precision, mode);
  return dval;
}

double HHVM_FUNCTION(deg2rad, double number) { return number / 180.0 * k_M_PI;}
double HHVM_FUNCTION(rad2deg, double number) { return number / k_M_PI * 180.0;}

String HHVM_FUNCTION(decbin, int64_t number) {
  return String(string_long_to_base(number, 2), AttachString);
}
String HHVM_FUNCTION(dechex, int64_t number) {
  return String(string_long_to_base(number, 16), AttachString);
}
String HHVM_FUNCTION(decoct, int64_t number) {
  return String(string_long_to_base(number, 8), AttachString);
}
Variant HHVM_FUNCTION(bindec, CVarRef binary_string) {
  String str = binary_string.toString();
  return string_base_to_numeric(str.data(), str.size(), 2);
}
Variant HHVM_FUNCTION(hexdec, CVarRef hex_string) {
  String str = hex_string.toString();
  return string_base_to_numeric(str.data(), str.size(), 16);
}
Variant HHVM_FUNCTION(octdec, CVarRef octal_string) {
  String str = octal_string.toString();
  return string_base_to_numeric(str.data(), str.size(), 8);
}

Variant HHVM_FUNCTION(base_convert, CVarRef number, int64_t frombase, int64_t tobase) {
  if (!string_validate_base(frombase)) {
    throw_invalid_argument("Invalid frombase: %" PRId64, frombase);
    return false;
  }
  if (!string_validate_base(tobase)) {
    throw_invalid_argument("Invalid tobase: %" PRId64, tobase);
    return false;
  }
  String str = number.toString();
  Variant v = string_base_to_numeric(str.data(), str.size(), frombase);
  return String(string_numeric_to_base(v, tobase), AttachString);
}

Variant HHVM_FUNCTION(pow, CVarRef base, CVarRef exp) {
  int64_t bint, eint;
  double bdbl, edbl;
  DataType bt = base.toNumeric(bint, bdbl, true);
  DataType et = exp.toNumeric(eint, edbl, true);
  if (bt == KindOfInt64 && et == KindOfInt64 && eint >= 0) {
    if (eint == 0) return 1LL;
    if (bint == 0) return 0LL;

    // calculate pow(long,long) in O(log exp) operations, bail if overflow
    int64_t l1 = 1;
    while (eint >= 1) {
      int overflow;
      double dval = 0.0;
      if (eint % 2) {
        --eint;
        ZEND_SIGNED_MULTIPLY_LONG(l1, bint, l1, dval, overflow);
        if (overflow) return dval * pow(bint, eint);
      } else {
        eint /= 2;
        ZEND_SIGNED_MULTIPLY_LONG(bint, bint, bint, dval, overflow);
        if (overflow) return (double)l1 * pow(dval, eint);
      }
      if (eint == 0) {
        return l1;
      }
    }
  }
  if (bt != KindOfDouble) {
    bdbl = base.toDouble();
  }
  if (et != KindOfDouble) {
    edbl = exp.toDouble();
  }
  return pow(bdbl, edbl);
}

double HHVM_FUNCTION(exp, double arg) { return exp(arg);}
double HHVM_FUNCTION(expm1, double arg) { return expm1(arg);}
double HHVM_FUNCTION(log10, double arg) { return log10(arg);}
double HHVM_FUNCTION(log1p, double number) { return log1p(number);}
double HHVM_FUNCTION(log, double arg, double base /* = 0 */) {
  return base <= 0 ? log(arg) : log(arg)/log(base);
}

double HHVM_FUNCTION(cos, double arg) { return cos(arg);  }
double HHVM_FUNCTION(cosh, double arg) { return cosh(arg); }
double HHVM_FUNCTION(sin, double arg) { return sin(arg);  }
double HHVM_FUNCTION(sinh, double arg) { return sinh(arg); }
double HHVM_FUNCTION(tan, double arg) { return tan(arg);  }
double HHVM_FUNCTION(tanh, double arg) { return tanh(arg); }
double HHVM_FUNCTION(acos, double arg) { return acos(arg); }
double HHVM_FUNCTION(acosh, double arg) { return acosh(arg);}
double HHVM_FUNCTION(asin, double arg) { return asin(arg); }
double HHVM_FUNCTION(asinh, double arg) { return asinh(arg);}
double HHVM_FUNCTION(atan, double arg) { return atan(arg); }
double HHVM_FUNCTION(atanh, double arg) { return atanh(arg);}
double HHVM_FUNCTION(atan2, double y, double x) { return atan2(y, x);}

double HHVM_FUNCTION(hypot, double x, double y) { return hypot(x, y);}
double HHVM_FUNCTION(fmod, double x, double y) { return fmod(x, y);}
double HHVM_FUNCTION(sqrt, double arg) { return sqrt(arg);}

int64_t HHVM_FUNCTION(getrandmax) { return RAND_MAX;}

///////////////////////////////////////////////////////////////////////////////

static bool s_rand_is_seeded = false;

void HHVM_FUNCTION(srand, CVarRef seed /* = null_variant */) {
  s_rand_is_seeded = true;
  if (seed.isNull()) {
    return srand(math_generate_seed());
  }
  if (seed.isNumeric(true)) {
    srand(seed.toInt32());
  } else {
    raise_warning("srand() expects parameter 1 to be long");
  }
}

TypedValue* HHVM_FUNCTION(rand, ActRec* ar) {
  TypedValue rvSpace = make_tv<KindOfInt64>(0);
  TypedValue* rv = &rvSpace;
  int32_t count = ar->numArgs();
  TypedValue* args UNUSED = ((TypedValue*)ar) - 1;
  if (LIKELY(count == 0 || count == 2)) {
    if (UNLIKELY(count == 2 && ((args - 0)->m_type != KindOfInt64
                                  || (args - 1)->m_type != KindOfInt64))) {
      if ((args - 0)->m_type != KindOfInt64) {
        tvCastToInt64InPlace(args-0);
      }
      if ((args - 1)->m_type != KindOfInt64) {
        tvCastToInt64InPlace(args-1);
      }
    }
    int64_t min, max;
    if (count == 0) {
      min = 0;
      max = HHVM_FN(getrandmax)();
    } else {
      min = args[-0].m_data.num;
      max = args[-1].m_data.num;
    }
    rvSpace.m_data.num = hphp_rand(min, max);
  } else {
    throw_wrong_arguments_nr("rand", count, 1, 2, 1, rv);
  }
  frame_free_locals_no_this_inl(ar, 2);
  ar->m_r = *rv;
  return &ar->m_r;
}

int64_t hphp_rand(int64_t min /* = 0 */, int64_t max /* = RAND_MAX */) {
  if (!s_rand_is_seeded) {
    s_rand_is_seeded = true;
    srand(math_generate_seed());
  }

  int64_t number = rand();
  if (min != 0 || max != RAND_MAX) {
    RAND_RANGE(number, min, max, RAND_MAX);
  }
  return number;
}

int64_t HHVM_FUNCTION(mt_getrandmax) { return MT_RAND_MAX;}

TypedValue* HHVM_FUNCTION(mt_rand, ActRec* ar) {
  TypedValue rvSpace = make_tv<KindOfInt64>(0);
  TypedValue* rv = &rvSpace;
  int32_t count = ar->numArgs();
  TypedValue* args UNUSED = ((TypedValue*)ar) - 1;
  if (LIKELY(count == 0 || count == 2)) {
    if (UNLIKELY(count == 2 && ((args - 0)->m_type != KindOfInt64
                                  || (args - 1)->m_type != KindOfInt64))) {
      if ((args - 0)->m_type != KindOfInt64) {
        tvCastToInt64InPlace(args-0);
      }
      if ((args - 1)->m_type != KindOfInt64) {
        tvCastToInt64InPlace(args-1);
      }
    }
    int64_t min, max;
    if (count == 0) {
      min = 0;
      max = HHVM_FN(mt_getrandmax)();
    } else {
      min = args[-0].m_data.num;
      max = args[-1].m_data.num;
    }
    rvSpace.m_data.num = hphp_mt_rand(min, max);
  } else {
    throw_wrong_arguments_nr("mt_rand", count, 1, 2, 1, rv);
  }
  frame_free_locals_no_this_inl(ar, 2);
  ar->m_r = *rv;
  return &ar->m_r;
}

void HHVM_FUNCTION(mt_srand, CVarRef seed /* = null_variant */) {
  if (seed.isNull()) {
    return math_mt_srand(math_generate_seed());
  }
  if (seed.isNumeric(true)) {
    math_mt_srand(seed.toInt32());
  } else {
    raise_warning("mt_srand() expects parameter 1 to be long");
  }
}

int64_t hphp_mt_rand(int64_t min /* = 0 */, int64_t max /* = MT_RAND_MAX */) {
  return math_mt_rand(min, max);
}
double HHVM_FUNCTION(lcg_value) { return math_combined_lcg();}

///////////////////////////////////////////////////////////////////////////////

const StaticString s_PHP_ROUND_HALF_UP("PHP_ROUND_HALF_UP");
const StaticString s_PHP_ROUND_HALF_DOWN("PHP_ROUND_HALF_DOWN");
const StaticString s_PHP_ROUND_HALF_EVEN("PHP_ROUND_HALF_EVEN");
const StaticString s_PHP_ROUND_HALF_ODD("PHP_ROUND_HALF_ODD");
const StaticString s_M_PI("M_PI");

class StandardMathExtension : public Extension {
 public:
  StandardMathExtension() : Extension("math") {}
  virtual void moduleInit() {
    Native::registerConstant<KindOfInt64>(
      s_PHP_ROUND_HALF_UP.get(), k_PHP_ROUND_HALF_UP
    );
    Native::registerConstant<KindOfInt64>(
      s_PHP_ROUND_HALF_DOWN.get(), k_PHP_ROUND_HALF_DOWN
    );
    Native::registerConstant<KindOfInt64>(
      s_PHP_ROUND_HALF_EVEN.get(), k_PHP_ROUND_HALF_EVEN
    );
    Native::registerConstant<KindOfInt64>(
      s_PHP_ROUND_HALF_ODD.get(), k_PHP_ROUND_HALF_ODD
    );
    Native::registerConstant<KindOfDouble>(
      s_M_PI.get(), k_M_PI
    );

    HHVM_FE(abs);
    HHVM_FE(is_finite);
    HHVM_FE(is_infinite);
    HHVM_FE(is_nan);
    HHVM_FE(ceil);
    HHVM_FE(floor);
    HHVM_FE(round);
    HHVM_FE(deg2rad);
    HHVM_FE(rad2deg);
    HHVM_FE(decbin);
    HHVM_FE(dechex);
    HHVM_FE(decoct);
    HHVM_FE(bindec);
    HHVM_FE(hexdec);
    HHVM_FE(octdec);
    HHVM_FE(base_convert);
    HHVM_FE(pow);
    HHVM_FE(exp);
    HHVM_FE(expm1);
    HHVM_FE(log10);
    HHVM_FE(log1p);
    HHVM_FE(log);
    HHVM_FE(cos);
    HHVM_FE(cosh);
    HHVM_FE(sin);
    HHVM_FE(sinh);
    HHVM_FE(tan);
    HHVM_FE(tanh);
    HHVM_FE(acos);
    HHVM_FE(acosh);
    HHVM_FE(asin);
    HHVM_FE(asinh);
    HHVM_FE(atan);
    HHVM_FE(atanh);
    HHVM_FE(atan2);
    HHVM_FE(hypot);
    HHVM_FE(fmod);
    HHVM_FE(sqrt);
    HHVM_FE(getrandmax);
    HHVM_FE(srand);
    HHVM_FE(rand);
    HHVM_FE(mt_getrandmax);
    HHVM_FE(mt_srand);
    HHVM_FE(mt_rand);
    HHVM_FE(lcg_value);

    loadSystemlib();
  }
} s_standardmath_extension;

}
